dali/pipeline/data/tensor_vector.cc

```diff
@@ -431,285 +666,354 @@ bool TensorVector<Backend>::is_pinned() const {
 
 
 template <typename Backend>
-int TensorVector<Backend>::device_id() const {
-  if (IsContiguous()) {
-    return tl_->device_id();
-  } else if (!tensors_.empty()) {
-    return tensors_[0]->device_id();
+void TensorVector<Backend>::set_device_id(int device_id) {
```

Comment:
Not really about this line - but related:
Now that we require a (hopefully valid?) device_id in share_data, I think that Buffer::set_device_id should throw if the pointer is not null.